src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,417 @@
+pub struct InputPortView {}
+impl ShapeViewDefinition for InputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(1.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Shape view for Input Port.
+#[derive(Debug,Clone,Copy)]
+pub struct OutputPortView {}
+impl ShapeViewDefinition for OutputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(0.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Port definition.
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct Port<T:ShapeViewDefinition+Clone> {
+    pub data : Rc<PortData<T>>
+}
+
+/// Type that represents an input port.
+pub type InputPort = Port<InputPortView>;
+
+/// Type that represents an output port.
+pub type OutputPort = Port<OutputPortView>;
+
+/// Internal data of `Port.
+#[derive(Debug,Clone)]
+#[allow(missing_docs)]
+pub struct PortData<T:ShapeViewDefinition+Clone> {
+        spec : RefCell<Specification>,
+    pub view : Rc<component::ShapeView<T>>
+}
+
+impl<T:ShapeViewDefinition<Shape=shape::Shape>+Clone> Port<T> {
+
+    /// Constructor.
+    pub fn new(spec:Specification) -> Self {
+        let logger = Logger::new("node");
+        let view   = Rc::new(component::ShapeView::new(&logger));
+        let spec = RefCell::new(spec);
+        let data = PortData{spec,view};
+        Self{data: Rc::new(data)}.init()
+    }
+
+    fn init(self) -> Self {
+        self.update();
+        self
+    }
+
+    /// Modifies the port specification.
+    pub fn mod_specification<F:FnOnce(&mut Specification)>(&mut self, f:F) {
+        f(&mut self.data.spec.borrow_mut());
+        self.update()
+    }
+
+    /// Update the shape parameters and sprite location with values from our current specification.
+    fn update(&self) {
+        self.update_shape();
+        self.update_sprite();
+    }
+
+    /// Update the position of the sprite according to the Port specification.
+    /// The position is given along a circle, thus the position and rotation of the sprite
+    /// are tied together, so the Port always point in the right direction.
+    fn update_sprite(&self) {
+        let spec = self.data.spec.borrow();
+        let translation_vector = nalgebra::Vector3::new(0.0,spec.inner_radius,0.0);
+        let rotation_vector    = -nalgebra::Vector3::new(0.0,0.0,spec.location.rad().value);
+        let rotation           = nalgebra::Rotation3::new(rotation_vector);
```

Comment:
This code does not seem linted. prelude exports `Vector3` so this qualification is not needed and linter should tell about it.